ql/src/java/org/apache/hadoop/hive/ql/scheduled/ScheduledQueryExecutionService.java

```diff
@@ -48,18 +53,29 @@
   private static ScheduledQueryExecutionService INSTANCE = null;
 
   private ScheduledQueryExecutionContext context;
-  private ScheduledQueryExecutor worker;
   private AtomicInteger forcedScheduleCheckCounter = new AtomicInteger();
+  private ScheduledQueryPoller poller;
+  AtomicInteger usedExecutors = new AtomicInteger(0);
+  List<ScheduledQueryExecutor> runningWorkers = new LinkedList<>();
 
   public static ScheduledQueryExecutionService startScheduledQueryExecutorService(HiveConf inputConf) {
     HiveConf conf = new HiveConf(inputConf);
     MetastoreBasedScheduledQueryService qService = new MetastoreBasedScheduledQueryService(conf);
-    ExecutorService executor = Executors.newCachedThreadPool(
-        new ThreadFactoryBuilder().setDaemon(true).setNameFormat("Scheduled Query Thread %d").build());
+    ExecutorService executor = buildExecutor(conf);
     ScheduledQueryExecutionContext ctx = new ScheduledQueryExecutionContext(executor, conf, qService);
     return startScheduledQueryExecutorService(ctx);
   }
 
+  private static ExecutorService buildExecutor(HiveConf conf) {
+    ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true).setNameFormat("Scheduled Query Thread %d").build();
+    int systemThreads = 2; // poller,reporter
+    int maxServiceThreads = conf.getIntVar(ConfVars.HIVE_SCHEDULED_QUERIES_MAX_EXECUTORS);
+    return new ThreadPoolExecutor(systemThreads + 1, systemThreads + maxServiceThreads,
```

Comment:
Can we leave a comment here? It seems we are always keeping at least poller, reporter, and an executor? It seems to me that minimum number of conf value HIVE_SCHEDULED_QUERIES_MAX_EXECUTORS should be 1 for this to work properly; though this is probably evident, we may leave a bit of this information to users in HiveConf description of the property.